### PR TITLE
Add support for Maven resources

### PR DESCRIPTION
**Pull request by LeoRamos01**
Generate "resource" tag inside pom.xml "build" tag.

### CODE REVIEW

File: initializr-generator/src/main/java/io/spring/initializr/generator/buildsystem/maven/MavenBuild.java [hunk 1]
```diff
@@ -134,4 +138,66 @@ public String getPackaging() {
...
+	 *
+	 * Builder to create a {@link Resource}.
+	 *
+	 *
+	 */
+	public static final class ResourceBuilder {
+
+		private String directory;
+
+		private List<String> includes = new ArrayList<>();
```
Comment (snicoll) [inline initializr-generator/src/main/java/io/spring/initializr/generator/buildsystem/maven/MavenBuild.java]: A maven resource has more than a directory and includes, see [the POM reference](https://maven.apache.org/pom.html#Resources).

File: initializr-generator/src/main/java/io/spring/initializr/generator/buildsystem/maven/MavenBuild.java [hunk 2]
```diff
@@ -134,4 +138,66 @@ public String getPackaging() {
...
+
+		public ResourceBuilder(String directory) {
+			this.directory = directory;
+		}
+
+		Resource build() {
+			return new Resource(this.directory, this.includes);
+		}
+
+		public ResourceBuilder include(String include) {
```
Comment (snicoll) [inline initializr-generator/src/main/java/io/spring/initializr/generator/buildsystem/maven/MavenBuild.java]: Looks like a missed opportunity to be able to specify multiple include patterns at once.

File: initializr-generator/src/main/java/io/spring/initializr/generator/buildsystem/maven/MavenBuild.java (hunk 1 above)
Comment (LeoRamos01) [reply initializr-generator/src/main/java/io/spring/initializr/generator/buildsystem/maven/MavenBuild.java]: Thanks! I'll make the needed changes as soon as possible!

**Comment by LeoRamos01** [comment]
> Thanks for the PR. Before we move forward with your proposal, tests have to be updated to exercise the code you've added. It is also very centric on a particular use case and need to be extended to properly cover the `resources` section of the pom.

I changed if you could have a look, thanks.

**Comment by LeoRamos01** [comment]
@snicoll Could you have a look at the updates, please?

**Comment by snicoll** [comment]
@LeoRamos01 please be patient, I am doing everything I can to catch up on things.

File: initializr-generator/src/test/java/io/spring/initializr/generator/buildsystem/maven/MavenBuildTests.java [hunk 3]
```diff
@@ -110,4 +114,29 @@ void mavenPluginExtensionsCanBeLoaded() {
 		assertThat(testPlugin.isExtensions()).isTrue();
 	}
 
+	@Test
+	void mavenResourcesCanBeLoaded() {
```
Comment (snicoll) [inline initializr-generator/src/test/java/io/spring/initializr/generator/buildsystem/maven/MavenBuildTests.java]: One test is not enough I am afraid, look at the other tests in this class for inspiration.

File: initializr-generator/src/main/java/io/spring/initializr/generator/buildsystem/maven/MavenBuild.java [hunk 4]
```diff
@@ -134,4 +138,110 @@ public String getPackaging() {
...
+
+		public ResourceBuilder(String directory) {
+			this.directory = directory;
+		}
+
+		Resource build() {
+			return new Resource(this.directory, this.targetPath, this.filtering, this.includes, this.excludes);
+		}
+
+		public ResourceBuilder include(List<String> includes) {
```
Comment (snicoll) [inline initializr-generator/src/main/java/io/spring/initializr/generator/buildsystem/maven/MavenBuild.java]: I meant specifying a vararg, not a list.

File: initializr-generator/src/main/java/io/spring/initializr/generator/buildsystem/maven/MavenBuild.java [hunk 5]
```diff
@@ -134,4 +138,110 @@ public String getPackaging() {
...
+			this.includes = includes;
+			return this;
+		}
+
+		public ResourceBuilder targetPath(String targetPath) {
+			this.targetPath = targetPath;
+			return this;
+		}
+
+		public ResourceBuilder filtering(Boolean filtering) {
```
Comment (snicoll) [inline initializr-generator/src/main/java/io/spring/initializr/generator/buildsystem/maven/MavenBuild.java]: I don't think it should use the wrapper type. The default should match what Maven does (`false`).

**Comment by snicoll** [comment]
@LeoRamos01 We've just pushed a `MavenPluginContainer` and I think the code here could use some harmonization in that direction. It has method to add/remove items (here `resource`)  and a customizer approach. The boundary between builder or not is still a bit vague at this point so please keep it as it is and we'll figure things out.

Do you have time to review your PR in that direction?

**Comment by LeoRamos01** [comment]
> @LeoRamos01 We've just pushed a `MavenPluginContainer` and I think the code here could use some harmonization in that direction. It has method to add/remove items (here `resource`) and a customizer approach. The boundary between builder or not is still a bit vague at this point so please keep it as it is and we'll figure things out.
> 
> Do you have time to review your PR in that direction?

Yes! As soon as possible I'll have a look!

File: initializr-generator/src/test/java/io/spring/initializr/generator/buildsystem/maven/MavenBuildTests.java (hunk 3 above)
Comment (LeoRamos01) [reply initializr-generator/src/test/java/io/spring/initializr/generator/buildsystem/maven/MavenBuildTests.java]: I've added some tests, see if its enough

**Comment by snicoll** [comment]
@LeoRamos01 I am afraid not. There aren't any test for `MavenBuildWriter`. We need tests that shows that writing such element to the pom happens as expected.

**Comment by LeoRamos01** [comment]
@snicoll Could you please help me?

I did not change the imports ordering and Spring is complaining about that.

I also don't understand this: **SpringLambda: Lambda block is unnecessary**.

**Comment by snicoll** [comment]
Thanks for your efforts @LeoRamos01. In the interesting of reducing the noise in notifications, please run the build locally and fix all issues before submitting a change (each commit triggers a notification).

`mvn spring-javaformat:apply` will format the code for you. That doesn't include import orders but you must have had a change as master builds cleanly now. For the lambda part, you need to wrap lambda parameters in `()`, i.e `(myparam) -> myparam.callMethod()`

**Comment by snicoll** [comment]
Alright, let me review what you've contributed.

**Comment by snicoll** [comment]
@LeoRamos01 thank you for making you first contribution to Spring Initializr. Please see [my polish commit](https://github.com/spring-io/initializr/commit/b711ac73c9b3b90d756a09266e2a3309040e2a1e) for what I've done based on our discussion.

**Pull request closed without merging**